File: backend/src/main/java/es/udc/paproject/backend/model/services/CatalogService.java
package es.udc.paproject.backend.model.services;

import es.udc.paproject.backend.model.entities.*;
import es.udc.paproject.backend.model.exceptions.DuplicateInstanceException;
import es.udc.paproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.paproject.backend.model.exceptions.OwnerOfProductException;
import es.udc.paproject.backend.model.exceptions.UserNotSellerException;

import java.util.List;
import java.util.Optional;

public interface CatalogService {


    /**
     *
     * @return All the crafts in alphabetical order
     */
    List<Craft> findAllCrafts();

    /**
     *
     * @return All the categories in alphabetical order and their subcategories
     */
    List<Category> findAllCategories();


    /**
     * Returns all active Products, that match the given criteria
     */
    Block<Product> findProducts(Long craftId, Long subcategoryId, String keywords, String productType, int page, int size);


    /**
     * Returns a specific Product that matches the productId
     */
    Product findProductById(Long productId) throws InstanceNotFoundException;

    void getProductType(Product product);

    /**
     * Returns all active products, whose owner matches the username.
     */
    Block<Product> findSellerProducts(String username, int page, int size) throws InstanceNotFoundException, UserNotSellerException;


    /**
     * Return all users with the seller role, whose username matches the given one.
     */
    Block<User> findSellers(String username, int page, int size);


    Favorite markAsFavoriteProduct(Long userId, Long productId) throws InstanceNotFoundException, DuplicateInstanceException, OwnerOfProductException;

    List<Product> getFavoriteProducts(Long userId) throws InstanceNotFoundException;

    Optional<Favorite> findFavoriteByUserAndProduct(Long userId, Long productId) throws InstanceNotFoundException;

    Favorite findFavoriteById(Long favoriteId) throws InstanceNotFoundException;

    void removeFavoriteProduct(Long userId, Long productId) throws InstanceNotFoundException;

}

File: frontend/src/backend/catalogService.js
import {appFetch, config} from "./appFetch.js";

export const findAllCrafts = (onSuccess) =>
    appFetch('/catalog/crafts' , config('GET'), onSuccess);


export const findAllCategories = (onSuccess) =>
    appFetch('/catalog/categories' , config('GET'), onSuccess);

export const findSubcategoriesByCategory = (categoryId, onSuccess)=>
    appFetch(`/catalog/${categoryId}/subcategories`, config('GET'), onSuccess);


export const findProducts = ({craftId, subcategoryId, keywords, productType, page}, onSuccess) => {

    let path=`/catalog/products?page=${page}`;

    path += craftId ? `&craftId=${craftId}`: "";
    path += subcategoryId ? `&subcategoryId=${subcategoryId}`: "";
    path += keywords.length > 0 ? `&keywords=${encodeURIComponent(keywords)}`: "";
    path += productType ? `&productType=${productType}`: "";

    appFetch(path, config('GET'), onSuccess);
}

export const findProductById = (id, onSuccess) =>
    appFetch(`/catalog/products/${id}`, config('GET'), onSuccess);

export const findUserProducts = (username,{page}, onSuccess, onErrors) => {
    appFetch(`/catalog/users/${username}/products?page=${page}`, config('GET'), onSuccess, onErrors);
}

export const findUsers = ({username, page}, onSuccess) =>{

    let path = `/catalog/users?page=${page}`;

    path += username ? `&username=${username}` : "";

    appFetch(path, config('GET'), onSuccess);
}